File: LeetCode_Python/slidingWindow/nSums/167.two-sum-ii-input-array-is-sorted.py
#
# @lc app=leetcode id=167 lang=python3
#
# [167] Two Sum II - Input Array Is Sorted
#

# @lc code=start
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        leftP = 0
        rightP = len(numbers)-1
        while (leftP < rightP):
            twoSum = numbers[leftP] + numbers[rightP]
            if(twoSum == target):
                return [leftP+1, rightP+1]
            if(twoSum < target):
                leftP+=1
            else:
                rightP-=1
        return [-1, -1]



# 在数组头和数组尾放置窗口此时窗口为最大
# 如果窗口和比target小，从数组头向右缩小窗口以增加sum      
# 如果窗口和比target小，从数组尾向左缩小窗口以减少sum       
# O(N)

# @lc code=end

File: LeetCode_Go/List&LinkedList/167.two-sum-ii-input-array-is-sorted.go
/*
 * @lc app=leetcode id=167 lang=golang
 *
 * [167] Two Sum II - Input Array Is Sorted
 */

// @lc code=start

// Given an sorted array of integers `nums` and an integer `target`, return indices of the two numbers such that they add up to target. Assume that each input would have exactly one solution. Elements in the array may not be unique.

// two pointers from both end, if sum larger then target, shrink the right, else shrink the left.

func twoSum(numbers []int, target int) []int {
	result := make([]int, 2)
	pLeft, pRight := 0, len(numbers)-1
	for pLeft <= pRight {
		if numbers[pLeft]+numbers[pRight] < target {
			pLeft++
		} else if numbers[pLeft]+numbers[pRight] > target {
			pRight--
		} else {
			result[0] = pLeft + 1
			result[1] = pRight + 1
			break
		}
	}
	return result
}

// @lc code=end

